casper/src/main/scala/coop/rchain/casper/api/GraphGenerator.scala

```diff
@@ -13,17 +13,143 @@ import cats.mtl._
 import cats.mtl.implicits._
 import coop.rchain.catscontrib.ski._
 
-case class Acc[G[_]](timeseries: List[Long] = List.empty, graph: G[Graphz[G]])
+case class ValidatorBlock(
+    blockHash: String,
+    parentsHashes: List[String],
+    justifications: List[String]
+)
+
+object ValidatorBlock {
+  implicit val validatorBlockMonoid: Monoid[ValidatorBlock] = new Monoid[ValidatorBlock] {
+    def empty: ValidatorBlock = ValidatorBlock("", List.empty[String], List.empty[String])
+    def combine(vb1: ValidatorBlock, vb2: ValidatorBlock): ValidatorBlock =
+      ValidatorBlock(
+        vb2.blockHash,
+        vb1.parentsHashes ++ vb2.parentsHashes,
+        vb1.justifications ++ vb2.justifications
+      )
+  }
+}
+
+case class GraphConfig(showJustificationLines: Boolean = false)
 
 object GraphzGenerator {
 
-  type ValidatorsBlocks = Map[Long, (String, List[String])]
-  case class Acc2[G[_]](
+  type ValidatorsBlocks = Map[Long, ValidatorBlock]
+
+  case class DagInfo[G[_]](
       validators: Map[String, ValidatorsBlocks] = Map.empty,
-      timeseries: List[Long] = List.empty,
-      graph: G[Graphz[G]]
+      timeseries: List[Long] = List.empty
   )
 
+  object DagInfo {
+    def empty[G[_]]: DagInfo[G] = DagInfo[G]()
+  }
+
+  def dagAsCluster[
+      F[_]: Monad: Sync: MultiParentCasperRef: Log: SafetyOracle: BlockStore,
+      G[_]: Monad: GraphSerializer
+  ](
+      topoSort: Vector[Vector[BlockHash]],
+      lastFinalizedBlockHash: String,
+      config: GraphConfig
+  ): F[G[Graphz[G]]] =
```

Comment:
As it is now, Graphz[G] contains both the data structure and one of the possible interpreters for it (which requires the GraphSerializer and the  G-effect).

I think it'd be better to separate the Graphz[G] into:
 - pure non-parametrized data structure
 - an G-parameterized inerpreter for it that is using the graph serializer

Then, this method could be `dagAsGraphz[F: ...](...) : F[Graphz]`, and its output could go to `serializeGraphs[F: GraphSerializer] : F[String]`. Or if we really want it, we could have it `serializeGraphs[G: GraphSerializer] : G[Unit]` and then run the G (under F?) to extract the result.

Reply:
I've seen your test/diagnostic util you've deleted from the other commit, and I thought: how do I even test that? My idea would be:

1. Have a data structure for the **abstract** graph (not the graphiviz-styling-oriented one Graphz is)
2. Generate an `Arbitrary` `Graph`, call it g
3. Interpret g into Graph-viz annotated structure using test code. Call it `expected`.
4. Interpret g into a BlockDag using test code. Call it `input`
5. (the 'when' part of the test:) Interpret `input` into Graph-viz using prod code, call it `actual`
6. Compare `expected` and `actual`.

Maybe we could even s/Graph-viz structure/String for a more precise and bigger-scoped test. But it'd be harder to maintain.

Just an idea, given that I've not seen actual tests for this :)

Reply:
1. regarding structure vs serializer

This is exactly design that I was aiming for, for two reasons:

a) this represents how graphviz is constucted. Using the DOT language you can add edges, node, configurations in randomly - arbitrarly. There is very little restrictions on how youy provide your graph. You can first define node, then edges. Or first edges, then nodes and then subrgraphs (or any other mixture for all that matter). 
a) secondly the idea originally was to allow bulding arbitrary HUGE graphs that will not blow up the memory.

Bottom line is that, somebody might fold over its own internal structure (potentially even on the fly, reading data from the stream/iterator) and building graph along the way. Your serializer might accumulate data (as the current `StringSerialzier` is doing) but it could also be a serialzier with very little memory footprint (one that for example drops everything to a file).

Initial graph generator (the one that is removed in this PR) was pushing that concept. You were foldLefting the structure and the graph was serializaed alon the way - it was still using the `StringSerialzier` but it could be replaced with `FileSerializer` whenever needed. 

The new generator (dagCluster) can not into account that "low memory footprint" feature into account because of how limited the graphviz is in generating "cluster-like" visualizations. Nevertheless it would be a bad idea to remove the feature (once its already here), because it is not being used. 

TLDR: The desing is as you see it to have ability to create a lazy graphs (all effects are closed over `G[_]` that has a small memory footprint (useage of serializer). In theory you could process a huge, tremendous graph, reading it from stream, and serialize it to a file on a very low memory consumption.

2. Regarding tests

a) testing graphz module

Now, you are mistaken that there are NO tests for the graphz module. Please see  `GraphzSpec` . Those tests are checking if the graphz module is propely rendering graphs. And it is covering quite a portion of its features.

b) testing visualize-dag functionality

What we are missing though is testing not the graphz module, but testing the usegae of graphz to visualize DAG.

I have an open ticket to add tests to it (RCHAIN-2801). Features requests are starting to grow over this `visualize-dag` functionality. It started as something that might or might not be used. Since it is being used, and there are features requests I will add tests to it. It will be based on my (VisalizeDagTest) that I've been using locally and accidently added to this PR.

c) property based testing

I am huge fun of property based testing. But that kind of testing only works if you have clearly defined properties. I would have to think if there are any that immediately are obvious. If you have an idea let me know.